src/pybind/mgr/dashboard/controllers/auth.py

```diff
@@ -28,29 +28,47 @@ class Auth(RESTController):
     """
     Provide authenticates and returns JWT token.
     """
-
     def create(self, username, password):
         user_data = AuthManager.authenticate(username, password)
         user_perms, pwd_expiration_date, pwd_update_required = None, None, None
-        if user_data:
-            user_perms = user_data.get('permissions')
-            pwd_expiration_date = user_data.get('pwdExpirationDate', None)
-            pwd_update_required = user_data.get('pwdUpdateRequired', False)
-
-        if user_perms is not None:
-            logger.debug('Login successful')
-            token = JwtManager.gen_token(username)
-            token = token.decode('utf-8')
-            cherrypy.response.headers['Authorization'] = "Bearer: {}".format(token)
-            return {
-                'token': token,
-                'username': username,
-                'permissions': user_perms,
-                'pwdExpirationDate': pwd_expiration_date,
-                'sso': mgr.SSO_DB.protocol == 'saml2',
-                'pwdUpdateRequired': pwd_update_required
-            }
+        attempt = mgr.ACCESS_CTRL_DB.get_attempt(username)
+        if attempt < 10:
+            if user_data:
+                user_perms = user_data.get('permissions')
+                pwd_expiration_date = user_data.get('pwdExpirationDate', None)
+                pwd_update_required = user_data.get('pwdUpdateRequired', False)
 
+            if user_perms is not None:
+                logger.debug('Login successful')
+                attempt = 0
+                mgr.ACCESS_CTRL_DB.save_attempt(username, attempt)
+                mgr.ACCESS_CTRL_DB.save()
+                token = JwtManager.gen_token(username)
+                token = token.decode('utf-8')
+                cherrypy.response.headers['Authorization'] = "Bearer: {}".format(token)
+                return {
+                    'token': token,
+                    'username': username,
+                    'permissions': user_perms,
+                    'pwdExpirationDate': pwd_expiration_date,
+                    'sso': mgr.SSO_DB.protocol == 'saml2',
+                    'pwdUpdateRequired': pwd_update_required
+                }
+            attempt += 1
+            mgr.ACCESS_CTRL_DB.save_attempt(username, attempt)
+            mgr.ACCESS_CTRL_DB.save()
+        else:
+            try:
+                user = mgr.ACCESS_CTRL_DB.get_user(username)
+                user.enabled = False
+                logger.debug('%s user is currently disabled', username)
+                raise DashboardException(msg='Invalid credentials',
+                                         code='invalid_credentials',
+                                         component='auth')
+            except UserDoesNotExist:
+                raise DashboardException(msg='Invalid credentials',
+                                         code='invalid_credentials',
+                                         component='auth')
```

Comment:
Instead of calling 3 times the same exception (with the risk of making a mistake/typo and allowing attackers distinguish the code path take here), I'd create an alias exception for this in the `..exceptions` file:
```py
InvalidCredentialsError = DashboardException(msg='Invalid credentials', code='invalid_credentials', component='auth')
...
raise InvalidCredentialsError
```

Reply:
I've created an alias and tried it. But the problem with that is the previously generated Exception gets stored into the InvalidCredentialError. So when an exception is raised the second time, it will also show the previous exception too. and for the third time, it'll show the previous two times.

Reply:
Right! So let's use subclassing instead:
```py
class InvalidCredentialsError(DashboardException):
  def __init__(self):
    super().__init__(msg='Invalid...)
```

Reply:
Works perfectly, thank you.